core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -422,11 +425,17 @@ private[log] class Cleaner(val id: Int,
       for (entry <- messages.shallowIterator) {
         val size = MessageSet.entrySize(entry.message)
         stats.readMessage(size)
-        if (entry.message.compressionCodec == NoCompressionCodec) {
+        val sourceCodec: CompressionCodec = entry.message.compressionCodec
+        val targetCodec = BrokerCompressionCodec.getTargetCompressionCodec(brokerCompressionType, sourceCodec)
+        if (sourceCodec == NoCompressionCodec) {
           if (shouldRetainMessage(source, map, retainDeletes, entry)) {
             val convertedMessage = entry.message.toFormatVersion(messageFormatVersion)
-            ByteBufferMessageSet.writeMessage(writeBuffer, convertedMessage, entry.offset)
-            stats.recopyMessage(size)
+            if (targetCodec == NoCompressionCodec) { // same as sourceCodec
```

Comment:
I think I'd write the condition as `sourceCodec == targetCode` and then you don't need the comment anymore.